src/rnnlm/rnnlm-simple-looped.h

```diff
@@ -0,0 +1,184 @@
+    opts->Register("frames-per-chunk", &frames_per_chunk,
+                   "Number of frames in each chunk that is separately evaluated "
+                   "by the neural net.");
+    opts->Register("debug-computation", &debug_computation, "If true, turn on "
+                   "debug for the actual computation (very verbose!)");
+    opts->Register("force-normalize", &force_normalize, "If true, force "
+                   " normalize the word posteriors");
+
+    // register the optimization options with the prefix "optimization".
+    ParseOptions optimization_opts("optimization", opts);
+    optimize_config.Register(&optimization_opts);
+
+    // register the compute options with the prefix "computation".
+    ParseOptions compute_opts("computation", opts);
+    compute_config.Register(&compute_opts);
+  }
+};
+
+class RnnlmSimpleLoopedInfo  {
+ public:
+  RnnlmSimpleLoopedInfo(
+      const RnnlmSimpleLoopedComputationOptions &opts,
+      const kaldi::nnet3::Nnet &rnnlm,
+      const CuMatrix<BaseFloat> &word_embedding_mat);
+
+  void Init(const RnnlmSimpleLoopedComputationOptions &opts,
+            const kaldi::nnet3::Nnet &rnnlm,
+            const CuMatrix<BaseFloat> &word_embedding_mat);
+
+  const RnnlmSimpleLoopedComputationOptions &opts;
+
+  const kaldi::nnet3::Nnet &rnnlm;
+  const CuMatrix<BaseFloat> &word_embedding_mat;
+
+  // frames_left_context equals the model left context plus the value of the
+  // --extra-left-context-initial option.
+  int32 frames_left_context;
+  // frames_right_context is the same as the right-context of the model.
+  int32 frames_right_context;
+  // The frames_per_chunk equals the number of input frames we need for each
+  // chunk (except for the first chunk).
+  int32 frames_per_chunk;
+
+  // The output dimension of the nnet neural network (not the final output).
+  int32 nnet_output_dim;
+
+  // The 3 computation requests that are used to create the looped
+  // computation are stored in the class, as we need them to work out
+  // exactly shich iVectors are needed.
+  ComputationRequest request1, request2, request3;
+
+  // The compiled, 'looped' computation.
+  NnetComputation computation;
+};
+
+/*
+  This class handles the neural net computation; it's mostly accessed
+  via other wrapper classes.
+
+  It accept just input features */
+class RnnlmSimpleLooped {
+ public:
+  /**
+     This constructor takes features as input.
+     Note: it stores references to all arguments to the constructor, so don't
+     delete them till this goes out of scope.
+
+     @param [in] info   This helper class contains all the static pre-computed information
+                        this class needs, and contains a pointer to the neural net.
+     @param [in] feats  The input feature matrix.
+  */
+  RnnlmSimpleLooped(const RnnlmSimpleLoopedInfo &info);
+
+  RnnlmSimpleLooped(const RnnlmSimpleLooped &other);
+
+  inline int32 NnetOutputDim() const { return info_.nnet_output_dim; }
```

Comment:
Rename to EmbeddingDim(), if needed, but I think this function should not be needed as this interface will just give you the log-prob of a next word.